polaris-service/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,172 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MetadataCacheManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetadataCacheManager.class);
+
+  /** Load the cached {@link Table} or fall back to `fallback` if one doesn't exist */
+  public static TableMetadata loadTableMetadata(
+      TableIdentifier tableIdentifier,
+      PolarisCallContext callContext,
+      PolarisEntityManager entityManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView,
+      Supplier<TableMetadata> fallback) {
+    LOGGER.debug(String.format("Loading cached metadata for %s", tableIdentifier));
+    Optional<TableMetadata> cachedMetadata =
+        loadCachedTableMetadata(tableIdentifier, callContext, entityManager, resolvedEntityView);
+    if (cachedMetadata.isPresent()) {
+      LOGGER.debug(String.format("Using cached metadata for %s", tableIdentifier));
+      return cachedMetadata.get();
+    } else {
+      TableMetadata metadata = fallback.get();
+      PolarisMetaStoreManager.EntityResult cacheResult =
+          cacheTableMetadata(
+              tableIdentifier, metadata, callContext, entityManager, resolvedEntityView);
+      if (!cacheResult.isSuccess()) {
+        LOGGER.debug(String.format("Failed to cache metadata for %s", tableIdentifier));
+      }
+      return metadata;
+    }
+  }
+
+  /**
+   * Attempt to add table metadata to the cache
+   *
+   * @return The result of trying to cache the metadata
+   */
+  private static PolarisMetaStoreManager.EntityResult cacheTableMetadata(
+      TableIdentifier tableIdentifier,
+      TableMetadata metadata,
+      PolarisCallContext callContext,
+      PolarisEntityManager entityManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView) {
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getPassthroughResolvedPath(tableIdentifier, PolarisEntitySubType.TABLE);
+    if (resolvedEntities == null) {
+      return new PolarisMetaStoreManager.EntityResult(
+          PolarisMetaStoreManager.ReturnStatus.ENTITY_NOT_FOUND, null);
+    } else {
+      TableLikeEntity tableEntity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+      TableMetadataEntity tableMetadataEntity =
+          new TableMetadataEntity.Builder()
+              .setCatalogId(tableEntity.getCatalogId())
+              .setParentId(tableEntity.getId())
+              .setId(entityManager.getMetaStoreManager().generateNewEntityId(callContext).getId())
+              .setCreateTimestamp(System.currentTimeMillis())
+              .setMetadataLocation(metadata.metadataFileLocation())
+              .setContent(TableMetadataParser.toJson(metadata))
+              .build();
+      try {
+        return entityManager
+            .getMetaStoreManager()
+            .createEntityIfNotExists(
+                callContext,
+                PolarisEntity.toCoreList(resolvedEntities.getRawFullPath()),
```

Comment:
Is this the first time we have something with a parentId that isn't a Namespace or a Catalog?

Reply:
`CATALOG_ROLE` is technically a child of a catalog, and we also have `FILE`